psycopg3/types/datetime.py

```diff
@@ -0,0 +1,112 @@
+"""
+Adapters for datetime types.
+https://www.postgresql.org/docs/9.4/datatype-datetime.html#DATATYPE-DATETIME-TABLE
+"""
+
+# Copyright (C) 2020 The Psycopg Team
+
+import codecs
+import datetime
+from typing import Union, Optional, Tuple
+
+from ..adapt import Dumper, Loader
+from ..proto import AdaptContext, DecodeFunc
+from .oids import builtins
+
+
+_encode = codecs.lookup("ascii").encode
+_decode = codecs.lookup("ascii").decode
+
+
+_time_oids = (
+    builtins["time"].oid,
+    builtins["timetz"].oid,
+    builtins["timestamp"].oid,
+    builtins["timestamptz"].oid,
+)
+
+
+_timezone_oids = (
+    builtins["timetz"].oid,
+    builtins["timestamptz"].oid,
+)
+
+
+@Dumper.text(datetime.date)
+@Dumper.binary(datetime.date)
+def dump_date(obj: datetime.date) -> Tuple[bytes, int]:
+    return _encode(str(obj))[0], builtins["date"].oid
+
+
+@Dumper.text(datetime.time)
+@Dumper.binary(datetime.time)
+def dump_time(obj: datetime.time) -> Tuple[bytes, int]:
+    if obj.tzinfo is None:
+        return _encode(str(obj))[0], builtins["time"].oid
+    else:
+        return _encode(str(obj))[0], builtins["timetz"].oid
+
+
+@Dumper.text(datetime.datetime)
+@Dumper.binary(datetime.datetime)
+def dump_datetime(obj: datetime.datetime) -> Tuple[bytes, int]:
+    if obj.tzinfo is None:
+        return _encode(str(obj))[0], builtins["timestamp"].oid
+    else:
+        return _encode(str(obj))[0], builtins["timestamptz"].oid
+
+
+@Dumper.text(datetime.timedelta)
+@Dumper.binary(datetime.timedelta)
+def dump_timedelta(obj: datetime.timedelta) -> Tuple[bytes, int]:
+    return _encode(str(obj))[0], builtins["interval"].oid
+
+
+@Loader.text(builtins["date"].oid)
+@Loader.binary(builtins["date"].oid)
+@Loader.text(builtins["time"].oid)
+@Loader.binary(builtins["time"].oid)
+@Loader.text(builtins["timetz"].oid)
+@Loader.binary(builtins["timetz"].oid)
+@Loader.text(builtins["timestamp"].oid)
+@Loader.binary(builtins["timestamp"].oid)
+@Loader.text(builtins["timestamptz"].oid)
+@Loader.binary(builtins["timestamptz"].oid)
```

Comment:
You can't use the same functions for binary and text decoding. I'm sure text format and binary format for dates are different.

Most likely these decorator will go each one on a different function, each one handling its own format.